packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.m

```diff
@@ -193,4 +208,34 @@ + (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
   return map;
 }
 
++ (SKPaymentDiscount *)getSKPaymentDiscountFromMap:(NSDictionary *)map {
+  if (!map || map.count <= 0) {
+    return nil;
+  }
+
+  NSAssert(map[@"identifier"],
+           @"When specifying a payment discount the 'identifier' field is mandatory.");
+  NSAssert(map[@"keyIdentifier"],
+           @"When specifying a payment discount the 'keyIdentifier' field is mandatory.");
+  NSAssert(map[@"nonce"], @"When specifying a payment discount the 'nonce' field is mandatory.");
+  NSAssert(map[@"signature"],
+           @"When specifying a payment discount the 'signature' field is mandatory.");
+  NSAssert(map[@"timestamp"],
+           @"When specifying a payment discount the 'timestamp' field is mandatory.");
+
+  NSString *identifier = map[@"identifier"];
+  NSString *keyIdentifier = map[@"keyIdentifier"];
+  NSUUID *nonce = [[NSUUID alloc] initWithUUIDString:map[@"nonce"]];
+  NSString *signature = map[@"signature"];
+  NSNumber *timestamp = map[@"timestamp"];
+
+  SKPaymentDiscount *discount = [[SKPaymentDiscount alloc] initWithIdentifier:identifier
```

Comment:
what happens to the `SKPaymentDiscount` object if any of the values are nil? Will it crash or throw some iap server error?

Reply:
If the `SKPaymentDiscount` object it self is `nil` the payment will continue without discount. This is also documented as part of the Dart API.

The values of a `SKPaymentDiscount` should never be `nil` as this would indicate an invalid discount offer. Apples documentation is not very clear on this but it should result into a server error. 

I have restructured the code and added additional validation to ensure empty and `nil` values are detected and returns a `PlatformError` in case invalid values are supplied. This to make it more clear to developers what went wrong.